[SPARK-28770][CORE][TEST] Fix ReplayListenerSuite tests that sometimes fail

wypoon opened the pull request:
### What changes were proposed in this pull request?

`ReplayListenerSuite` depends on a listener class to listen for replayed events. This class was implemented by extending `EventLoggingListener`. `EventLoggingListener` does not log executor metrics update events, but uses them to update internal state; on a stage completion event, it then logs stage executor metrics events using this internal state. As executor metrics update events do not get written to the event log, they do not get replayed. The internal state of the replay listener can therefore be different from the original listener, leading to different stage completion events being logged.

We reimplement the replay listener to simply buffer each and every event it receives. This makes it a simpler yet better tool for verifying the events that get sent through the ReplayListenerBus.

### Why are the changes needed?

As explained above. Tests sometimes fail due to events being received by the `EventLoggingListener` that do not get logged (and thus do not get replayed) but influence other events that get logged.

### Does this PR introduce any user-facing change?

No.

### How was this patch tested?

Existing unit tests.

wypoon committed:
[SPARK-28770][CORE][TEST] Fix ReplayListenerSuite tests that sometimes fail.

testApplicationReplay fails if the application runs long enough for the driver
to send an executor metrics update. This causes stage executor metrics to be
written for the driver. However, executor metrics updates are not logged, and
thus not replayed. Therefore no stage executor metrics for the driver is logged
on replay.

SparkQA commented:
**[Test build #110092 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110092/testReport)** for PR 25673 at commit [`0b0adb0`](https://github.com/apache/spark/commit/0b0adb0b7a508f3e1902b838382e478a1435cfde).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/15108/
Test PASSed.

HyukjinKwon commented on core/src/test/scala/org/apache/spark/scheduler/ReplayListenerSuite.scala:
```diff
@@ -219,11 +219,25 @@ class ReplayListenerSuite extends SparkFunSuite with BeforeAndAfter with LocalSp
+    // For this reason, exclude stage executor metrics for the driver.
+    val filteredEvents = originalEvents.filter { e =>
+      if (e.isInstanceOf[SparkListenerStageExecutorMetrics] &&
+        e.asInstanceOf[SparkListenerStageExecutorMetrics].execId == "driver") {
```
What I am not understanding is that, why are both different? They make JSONs from the same instances. Are they different? Why is it flaky?

HeartSaVioR replied:
EventMonster extends EventLoggingListener which doesn't transparently log events, hence some possible cases could exist where both can be different.

wypoon replied:
I tried to explain briefly in the comment. Let me try to explain in more detail.

When the application is run, executor metrics update events may occur (if the application runs long enough). These events are not written to the event log, so they do not get replayed. That is the root cause of differences.
On executor metrics update, the EventLoggingListener updates a map it uses to track per-stage metrics, but does not log. The listener also receives metrics on task end. Because of this, it always has metrics for executors. On stage completion. the listener logs stage executor metrics, for any of the executors/driver it has metrics for. As explained, sometimes it will have metrics for the driver (if an executor metrics update arrived from the driver), but most of the time not.
To recap, SparkListenerStageExecutorMetrics events are not events received while the application is running, but secondary events derived from metrics received and written to the event log. (One other point not relevant to this bug but may help understanding - on replay, the SparkListenerStageExecutorMetrics events do get replayed but are ignored by the listener.)

wypoon replied:
I had written the same explanation in https://github.com/apache/spark/pull/25659#issuecomment-527741025.

HeartSaVioR commented on core/src/test/scala/org/apache/spark/scheduler/ReplayListenerSuite.scala:
```diff
@@ -219,11 +219,25 @@ class ReplayListenerSuite extends SparkFunSuite with BeforeAndAfter with LocalSp
+    // (at heartbeat intervals), and when we do, it is only in the original events, never
+    // in the replay, since executor metrics updates are not replayed.
+    // For this reason, exclude stage executor metrics for the driver.
+    val filteredEvents = originalEvents.filter { e =>
```
I guess below would be simpler:

```
val filteredEvents = originalEvents.filter {
  case e: SparkListenerStageExecutorMetrics if e.execId == "driver" => false
  case _ => true
}
```

wypoon replied:
Yes, that's nicer. Thanks.

SparkQA commented:
**[Test build #110092 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110092/testReport)** for PR 25673 at commit [`0b0adb0`](https://github.com/apache/spark/commit/0b0adb0b7a508f3e1902b838382e478a1435cfde).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/110092/
Test PASSed.

HeartSaVioR commented:
I'd rather say we may need to find the way to let EventMonster not extending EventLoggingListener and just buffering all events. We don't need to deal with such complicated case in EventLoggingListener, as what we want to verify in ReplayListenerSuite is "ReplayListener can read all events sequentially and pass to registered listeners correctly".

https://github.com/apache/spark/blob/df39855db826fd4bead85a2ca01eda15c101bbbe/core/src/test/scala/org/apache/spark/scheduler/ReplayListenerSuite.scala#L236-L253

I'm a bit concerned about the javadoc of EventMonster - I expect each event are posted to be same to all listeners. If that's the understanding of us (Spark community), it would end up with inconsistency if we allow the event itself or any fields in event to be modified somewhere (e.g. #25672).

wypoon commented:
> I'd rather say we may need to find the way to let EventMonster not extending EventLoggingListener and just buffering all events. We don't need to deal with such complicated case in EventLoggingListener, as the purpose of ReplayListenerSuite is to test the fact that "ReplayListener can read all events sequentially and pass to registered listeners correctly".
> 
I'd rather not take on reworking EventMonster here. I think it is sufficient to handle the special case of `SparkListenerStageExecutorMetrics` by filtering.

wypoon committed:
[SPARK-28770][CORE][TEST] Simplify the filter.

Suggested by Jungtaek Lim.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/15144/
Test PASSed.

SparkQA commented:
**[Test build #110135 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110135/testReport)** for PR 25673 at commit [`7ba6b96`](https://github.com/apache/spark/commit/7ba6b969bc08e5a1d929c25d72f809647222af73).

HeartSaVioR commented:
The test is going to be unnecessary complicated, as I commented earlier. We are required to know about detailed behavior of EventLoggingListener, though we are in ReplayListenerSuite. ReplayListener is even not used with EventLoggingListener so it's really only for testing which doesn't make sense to couple with. Even we fix it for this time, the test might be broken again if we tune EventLoggingListener once more.

For sure, it's not your fault. You did a great analysis. The code looks to be ancient, so it's not their fault too. We just seem to miss the relevant codes while EventLoggingListener is evolving. But at least now we indicate the unnecessary complexity, it's your bet to just add band-aid or fix it on the right way. I prefer latter, and I'm OK if committers are OK with former and merge the patch as it is. Maybe I'll raise another PR to fix it then.

SparkQA commented:
**[Test build #110135 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110135/testReport)** for PR 25673 at commit [`7ba6b96`](https://github.com/apache/spark/commit/7ba6b969bc08e5a1d929c25d72f809647222af73).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/110135/
Test PASSed.

wypoon committed:
[SPARK-28770][CORE][TEST] Change implementation of EventMonster.

Instead of having EventMonster extend EventLoggingListener, have it extend
SparkFirehoseListener and simply buffer all events it receives. This makes
it much more suitable for verifying that the ReplyListnerBus replays all
the events from event logs.
With this new EventMonster, there is no need to have special exception
handling for certain types of events.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/15155/
Test PASSed.

SparkQA commented:
**[Test build #110149 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110149/testReport)** for PR 25673 at commit [`6fab21f`](https://github.com/apache/spark/commit/6fab21f447ee58cdd5b16c8e9cecc57061209747).

wypoon commented:
@HeartSaVioR, on second thoughts, after offline discussion with @squito, I agree that the best way to fix the tests is to reimplement EventMonster. Instead of extending EventLoggingListener, EventMonster should simply buffer any and all events it receives. @squito pointed out that EventMonster was added when ReplayListenerSuite was first added, and that SparkFirehoseListener did not exist then. Extending EventLoggingListener and adding a mode where it simply appends events to a buffer (in its logEvent method) was an easy way of implementing EventMonster. However, as EventLoggingListener evolved, it did not call logEvent on every event, and it also wrote events as side effects that depended on its internal state, leading to the current difficulties.
I have reimplemented EventMonster (extending SparkFirehoseListener) to simply buffer each and every event it receives. This matches the needs of ReplayListenerSuite better. With this change, testApplicationReplay no longer needs any logic to filter any events.

HeartSaVioR reviewed (approved):
LGTM

SparkQA commented:
**[Test build #110149 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110149/testReport)** for PR 25673 at commit [`6fab21f`](https://github.com/apache/spark/commit/6fab21f447ee58cdd5b16c8e9cecc57061209747).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/110149/
Test PASSed.

HeartSaVioR commented:
cc. @squito as he commented and discussed the approach of resolution in JIRA issue.

srowen commented on core/src/test/scala/org/apache/spark/scheduler/ReplayListenerSuite.scala:
```diff
@@ -236,20 +238,15 @@ class ReplayListenerSuite extends SparkFunSuite with BeforeAndAfter with LocalSp
    */
-  private class EventMonster(conf: SparkConf)
-    extends EventLoggingListener("test", None, new URI("testdir"), conf) {
+  private class EventMonster extends SparkFirehoseListener {
```
No need to change it here, but, Event_Monster_?

squito replied:
lol, good point.  EventBufferingListener would be better

wypoon replied:
Sure, I will rename it EventBufferingListener.

wypoon replied:
Updated the PR description as well to not mention EventMonster.

squito commented:
lgtm.  Also I see @bzhaoopenstack confirmed this does the trick on https://github.com/apache/spark/pull/25659.

wing yew, can you do the rename?  and update the PR descrpition to not mention EventMonster by name anymore

wypoon committed:
[SPARK-28770][CORE][TEST] Rename EventMonster to EventBufferingListener.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/15200/
Test PASSed.

SparkQA commented:
**[Test build #110194 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110194/testReport)** for PR 25673 at commit [`a0287a8`](https://github.com/apache/spark/commit/a0287a80e6c457efd14118a790d0d2be35a05ba4).

SparkQA commented:
**[Test build #110194 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110194/testReport)** for PR 25673 at commit [`a0287a8`](https://github.com/apache/spark/commit/a0287a80e6c457efd14118a790d0d2be35a05ba4).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/110194/
Test PASSed.

squito commented:
merged to master, thanks everyone
